### ISSUE TITLE
Question: is it expected that inner function objects include braces?

### ISSUE THREAD

**Issue by telemachus** [nvim-treesitter/nvim-treesitter-textobjects#112]
(Thanks for the plugin: it is a snap to set up and easy to work with so far.)

I was surprised to see that by default, an *inner* function object in, for example, Go and Javascript includes the braces. I'm guessing that this is expected behavior and probably a result of how treesitter itself works. But I wanted to confirm that before going further.

I ask because my most common use for inner function objects is that I want to change, delete, or yank all the code in the function body, but in those cases I don't normally want to include the braces. So, I'll need to consider whether it's possible to tweak treesitter's results.

**Comment by stsewd** [comment]
They shouldn't include braces. Depending on the grammar this could be easy to fix or not... For js and go, looks like it require for us to support the `*/+` operators.

```
(function_declaration
  body: (statement_block . "{" (_)* @function.inner . "}")) @function.outer
```

Could also be fixed by the grammar to not include braces inside the block node.

**Comment by stsewd** [comment]
Looks like we have the same problem with C :/

**Comment by dominikh** [comment]
I don't think that linewise makes sense for `ap`, since functions can be expressions in many languages, like in the following snippet of javascript:

```js
var x = function() { 
  console.log("Hello, world")
}
```

I do agree that `af` (and other `a`s) should include whitespace, however.

**Comment by theHamsta** [comment]
It is a short-coming that we can only select nodes at the moment. We wanted to trim the braces using a regex directive and use that solution also for language injection. Language injection now uses a hack `#offset` to trim characters. Would be great when someone could contribute a proper solution!

**Comment by Melkster** [comment]
@dominikh 
> I don't think that linewise makes sense for `ap`, since functions can be expressions in many languages

I'm not sure that I'm following, could you please elaborate? Also, did you mean to write `ap`, or should it be `if`?

**Comment by dominikh** [comment]
@Melkster 
I meant to write `af`, not `ap`. And my point was that if `af` acted linewise, then doing something like `daf` inside the function in 

```
var x = function() { 
  console.log("Hello, world")
}
```

would delete more than just the function.

**Comment by Melkster** [comment]
Perhaps there should be an additional command for a linewise "around function" text-object, for instance `Af`, kinda like how [targets.vim](https://github.com/wellle/targets.vim#around-pair) allows an upper-case "around" to do things like `A(`, `A{`, which includes any trailing whitespace.

I personally think that it would be a super useful feature to be able to move around functions using a single text-object which gets the whitespace right immediately, just like one can do with paragraphs and `ap`.

**Comment by telemachus** [comment]
> [M]y point was that if af acted linewise, then doing something like daf inside the function in
>
> ```
> var x = function() { 
> console.log("Hello, world")
> }
> ```
> would delete more than just the function.

@dominikh At first thought, I think that `daf` *should* delete more than the function there. If someone wanted only to rewrite the content of the anonymous function, they would use `cif`, right?

That said, I suppose that you're leaving room for cases where people want to replace the entire anonymous function and assign `x` to a named function or some other value?

For whatever it's worth, I'm back to using [the object mappings in vim-go](https://github.com/fatih/vim-go/blob/master/autoload/go/textobj.vim) (which rely on fatih's [motion](https://github.com/fatih/motion)), and `vaf` selects from `sayHelloWorld` to the closing `}` in the Go code below. I think I understand what you want, but this result basically matches my expectations. (There are issues there too though: if the cursor is on `sayHelloWorld`, vim-go does not see that line as the beginning of a function. `vaf` selects the entire function that `sayHelloWorld` is in. Similarly, if you use `var sayHelloWorld = func…` at package level, `vaf` with the cursor on `var` or `sayHelloWorld` is a no-op. Whether at package level or within a function, selection of anonymous functions assigned to variables only works once the cursor is at `func` on the first line or later.)

```go
	sayHelloWorld := func(userName string) {
		fmt.Println("Hello", userName)
	}
```

**Comment by theHamsta** [comment]
`daf` should really only delete the actual function. To delete line-wise you can use `dVaf`. Maybe one could have an option for users to make certain mappings automaticly line-wise.

**Comment by arsham** [comment]
I'm trying to exclude the braces for `go` based on the above suggestion, the it didn't match the inner function (or block of code actually). I've tried these variations:
```
(function_declaration
  body: (block . "{" (_)* @function.inner . "}")) @function.outer
```
And
```
(func_literal
  body: (block . "{" (_)* @function.inner . "}")) @function.outer
```

I have the latest Neovim and all related plugins in my setup.

Thank you.

**Comment by telemachus** [comment]
@arsham I haven't found a solution yet with this repository, but you may find [nvim-treesitter-textsubjects](https://github.com/RRethy/nvim-treesitter-textsubjects) helpful. (It only works for a handful of languages so far, but it's inner object works well with go.)

**Comment by telemachus** [comment]
@stsewd @theHamsta until `*` and `+` are supported for captures, would you consider a patch that uses the pattern from [nvim-treesitter-textsubjects](https://github.com/RRethy/nvim-treesitter-textsubjects) for inner functions? It definitely seems like the consensus that an inner function text object should not include the braces. The following work well for go, for example. (I think I should be able to write this as one query using alteration, but I can't get it right.)

```query
(function_declaration
  (block . "{" . (_) @_start @_end (_)? @_end . "}"
         (#make-range! "myfunction.inner" @_start @_end)))

(method_declaration
  (block . "{" . (_) @_start @_end (_)? @_end . "}"
         (#make-range! "myfunction.inner" @_start @_end)))

(func_literal
  (block . "{" . (_) @_start @_end (_)? @_end . "}"
         (#make-range! "myfunction.inner" @_start @_end)))
```

**Comment by theHamsta** [comment]
Would be ok. If possible all "{" should be patched

**Comment by telemachus** [comment]
> If possible all "{" should be patched

Yup, understood. I'm not sure how many of those languages I would immediately recognize as such, but I can do several, start a PR, and then get feedback about others to fill in. I'm happy to do that if it works for people.

**Comment by jjant** [comment]
This is still broken/wrong for `@function.inner` in Rust, and `@block.inner`. Both include the curly braces.

**Comment by telemachus** [comment]
> This is still broken/wrong for `@function.inner` in Rust, and `@block.inner`. Both include the curly braces.

Yup. Unfortunately [as I mentioned soon after](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/196#issuecomment-1102540930), I didn't think to include Rust. I don't use the language, and I didn't realize that it was affected when I first made the pull request.

There are [other ways in which my initial pull request probably needs to be expanded](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/196#issuecomment-1102355935). I meant to do that myself, but I frankly forgot.

I definitely won't get to this quickly (I'm a teacher, and it's recommendation-writing season), but by looking at [the changes I made to other languages](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/196/files), someone else could probably fix Rust pretty quickly. (You can also test it locally on your machine if you follow the instructions to [override predefined textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects#overriding-or-extending-textobjects).)

**Comment by jjant** [comment]
Ah, I see, I missed your comment!
Thanks for the pointers, I'll give it a try myself (if/when I've got the time 😅).

**Comment by icyd** [comment]
This PR #314 should solve this issue

**Comment by Goxore** [comment]
I have the same issue with c#

**Comment by icyd** [comment]
> I have the same issue with c#

I don't use C#, please verify this work for you #318

**Comment by ilan-schemoul** [comment]
I do not know if it's related but in c/cpp @block.inner includes braces #589
